test/base/src/androidTest/kotlin/io/realm/kotlin/test/shared/RealmSetTests.kt

```diff
@@ -0,0 +1,1137 @@
+        errorCatcher {
+            val dataSet = typeSafetyManager.dataSetToLoad
+            realm.writeBlocking {
+                typeSafetyManager.createContainerAndGetCollection(this)
+                    .addAll(dataSet)
+            }
+
+            val set = realm.query<RealmSetContainer>()
+                .first()
+                .find { setContainer ->
+                    assertNotNull(setContainer)
+                    typeSafetyManager.getCollection(setContainer)
+                }
+
+            realm.close()
+
+            assertFailsWith<IllegalStateException> {
+                set.iterator()
+            }
+            assertFailsWith<IllegalStateException> {
+                set.iterator().hasNext()
+            }
+            assertFailsWith<IllegalStateException> {
+                set.iterator().next()
+            }
+            assertFailsWith<IllegalStateException> {
+                set.iterator().remove()
+            }
+        }
+    }
+
+    override fun assertContainerAndCleanup(assertion: (RealmSetContainer) -> Unit) {
+        val container = realm.query<RealmSetContainer>()
+            .first()
+            .find()
+        assertNotNull(container)
+
+        // Assert
+        errorCatcher {
+            assertion(container)
+        }
+
+        // Clean up
+        realm.writeBlocking {
+            delete(findLatest(container)!!)
+        }
+    }
+}
+
+/**
+ * Tester for generic types.
+ */
+internal class GenericSetTester<T>(
+    realm: Realm,
+    typeSafetyManager: SetTypeSafetyManager<T>,
+    private val classifier: KClassifier
+) : ManagedSetTester<T>(realm, typeSafetyManager) {
+
+    override fun toString(): String = classifier.toString()
+
+    override fun assertStructuralEquality(expected: Collection<T>, actual: Collection<T>) {
+        assertEquals(expected.size, actual.size)
+        actual.forEach {
+            assertTrue(expected.contains(it))
+        }
+    }
+
+    override fun structuralContains(collection: Collection<T>, element: T?): Boolean =
+        collection.contains(element)
+}
+
+/**
+ * Tester for ByteArray.
+ */
+internal class ByteArraySetTester(
+    realm: Realm,
+    typeSafetyManager: SetTypeSafetyManager<ByteArray>,
+    private val classifier: KClassifier
+) : ManagedSetTester<ByteArray>(realm, typeSafetyManager) {
+
+    override fun toString(): String = classifier.toString()
+
+    override fun assertStructuralEquality(
+        expected: Collection<ByteArray>,
+        actual: Collection<ByteArray>
+    ) {
+        assertEquals(expected.size, actual.size)
+
+        // We can't iterate by index on the set and the positions are not guaranteed to be the same
+        // as in the dataset so to compare the values are the same we need to bend over backwards...
+        var assertionSucceeded = 0
+        actual.forEach { actualByteArray ->
+            expected.forEach { expectedByteArray ->
+                try {
```

Comment:
A bit convoluted. I guess it is overall sound as you verify that size of the two collections are equal in size up front, but would be more clear to iterate and verify that all elements of `actual` is actually in `expected` and break if it isn't. Then you could also print the element not found.